File: libglnsvpos/src/diffs.cpp
#include <cmath>
#include <include\libglnsvpos\diffs.h>

void dif(double *koord, double *diffs)
{
    double J_2 = 1082625.75e-9; //зональный гармонический коэффициент второй степени, характеризующий полярное сжатие Земли
    double GM = 398600.4418e9; //геоцентрическая константа гравитационного поля Земли с учетом атмосферы
    double a_e = 6378136; //большая полуось общеземного эллипсоида
    double r = sqrt(koord[0]*koord[0] + koord[1]*koord[1] + koord[2]*koord[2]);
    diffs[0] = koord[3];
    diffs[1] = koord[4];
    diffs[2] = koord[5];
    double GM0 = GM/(r*r);
    double x0 = koord[0]/r;
    double y0 = koord[1]/r;
    double z0 = koord[2]/r;
    double R = a_e/r;
    diffs[3] = -GM0*x0 - 1.5*J_2*GM0*x0*R*R*(1-5*z0*z0);
    diffs[4] = -GM0*y0 - 1.5*J_2*GM0*y0*R*R*(1-5*z0*z0);
    diffs[5] = -GM0*z0 - 1.5*J_2*GM0*z0*R*R*(3-5*z0*z0);
}

File: diffs.m
function dF = diffs(t, F)
%константы
GM = 398600441.8e6;
J02 = -1082.63*10^-6;
ae = 6378136;
%расчет
Xa=F(1);
Ya=F(2);
Za=F(3);
r=sqrt(Xa^2 + Ya^2 + Za^2);
GMn = GM/r^2;
x1 = Xa/r;
y1 = Ya/r;
z1 = Za/r;
RO = ae/r;
dF = F(:);
dF(1) = F(4);
dF(2) = F(5);
dF(3) = F(6); 
dF(4) = -GMn*x1 + 1.5*J02*GMn*x1*(RO^2)*(1 - 5*z1^2);
dF(5) = -GMn*y1 + 1.5*J02*GMn*y1*(RO^2)*(1 - 5*z1^2);
dF(6) = -GMn*z1 + 1.5*J02*GMn*z1*(RO^2)*(3 - 5*z1^2);
end

